### PR TITLE
KOA-5827 Migrate text token generation to kotlin

### PR DESCRIPTION
**Pull request by marianeum**
+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] Component `README.md`
+ [x] Tests
+ [x] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: Backpack/src/main/res/values/backpack.animation.xml [hunk 1]
```diff
@@ -17,6 +17,9 @@
   limitations under the License.
 -->
 <resources>
+
+  <!-- THIS FILE IS AUTO GENERATED. DO NOT MODIFY! -->
```
Comment (marianeum) [inline Backpack/src/main/res/values/backpack.animation.xml:21]: this was in the text file, but not any others - thought i'd include it in the template for all to make it more obvious what's autogenerated vs not

File: Backpack/src/main/res/values/backpack.text.size.xml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+<resources>
+
+  <!-- THIS FILE IS AUTO GENERATED. DO NOT MODIFY! -->
+
+  <dimen name="bpkTextXlSize">24sp</dimen>
```
Comment (marianeum) [inline Backpack/src/main/res/values/backpack.text.size.xml:23]: it was a looot simpler to split out the text sizes into a separate file. with xml resources the file doesn't matter to consumers, so it's not a breaking change

File: Backpack/src/main/res/values/backpack.text.xml [hunk 3]
```diff
@@ -20,63 +20,26 @@
...
-    <item name="android:fontFamily">?bpkFontFamilyEmphasized</item>
-    <item name="android:textColor">@color/bpkTextPrimary</item>
-    <item name="android:textSize">@dimen/bpkText6XlSize</item>
-    <item name="lineHeight">84sp</item>
-    <item name="android:letterSpacing">-0.02</item>
-  </style>
-
-  <style name="bpkTextFootnote">
+  
+  <style name="bpkTextBodyDefault">
```
Comment (marianeum) [inline Backpack/src/main/res/values/backpack.text.xml:31]: the order is different, but they're the same. the screenshot tests will validate that

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 4]
```diff
@@ -30,10 +30,15 @@ interface BpkTextStyles : List<BpkTextStyleModel>
 
 data class BpkTextStyleModel(
   val name: String,
-  val fontSize: String,
-  val fontWeight: String,
-  val lineHeight: String,
-  val letterSpacing: String?,
+  val fontSize: Token,
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:33]: with the xml tokens we sometimes need the name and sometimes the actual value, so had to change the parsing a little

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 5]
```diff
@@ -59,19 +69,15 @@ private fun parseTextStyles(source: Map<String, Any>): BpkTextStyles {
   fun String.trimReference(): String =
     removePrefix("{!").removeSuffix("}")
 
-  fun String.toSemanticName() =
-    CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, this.replace('-', '_'))
-
-  fun String.wrapIfDigits() = if (first().isDigit()) { "`$this`" } else this
-
-  fun List<Map.Entry<String, Map<String, String>>>.findTokenName(type: String): String? {
+  fun List<Map.Entry<String, Map<String, String>>>.findToken(type: String): Token? {
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:72]: also here had to change the parsing and remove the case conversion/wrapping logic and move that into the formatter later on

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 6]
```diff
@@ -88,10 +94,10 @@ private fun parseTextStyles(source: Map<String, Any>): BpkTextStyles {
     .map {
       BpkTextStyleModel(
         name = it.key,
-        fontSize = it.value.findTokenName("FONT_SIZE")!!,
-        fontWeight = it.value.findTokenName("FONT_WEIGHT")!!.let { if (it == "Book") "Normal" else it },
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:92]: the names are different in view as well, so moved that to the formatter

File: buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt [hunk 7]
```diff
@@ -134,23 +147,29 @@ private fun toCompose(source: BpkTextStyles, className: String): TypeSpec {
...
-        ) {
-          "${it.name.toSemanticName()} = ${it.toParameter()}"
-        }, mapOf(
-          "textStyle" to TextStyleClass,
-          "fontWeight" to FontWeightClass,
-          "fontSize" to BpkFontSizeClass,
-          "lineHeight" to BpkLineHeightClass,
-          "letterSpacing" to BpkLetterSpacingClass,
-        )).build())
+        .addParameter(
```
Comment (marianeum) [inline buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt:150]: this was just a formatting change that happened when i reformatted the file. not sure why detekt didn't pick up changes...

**Review by frugoman — APPROVED** [review]

**Pull request merged**